python/cufinufft/cufinufft/_plan.py

```diff
@@ -0,0 +1,373 @@
+    the execution of an NUFFT. First, the plan is created with a certain set
+    of parameters (type, mode configuration, tolerance, sign, number of
+    simultaneous transforms, and so on). Then the nonuniform points are set
+    (source or target depending on the type). Finally, the plan is executed on
+    some data, yielding the desired output.
+
+    Args:
+        nufft_type      (int): type of NUFFT (1 or 2).
+        n_modes         (tuple of ints): the number of modes in each
+                        dimension (for example `(50, 100)`).
+        n_trans         (int, optional): number of transforms to compute.
+        eps             (float, optional): precision requested (>1e-16).
+        isign           (int, optional): if +1, uses the positive sign
+                        exponential, otherwise the negative sign; defaults to
+                        +1 for type 1 and to -1 for type 2.
+        dtype           (string, optional): the precision of the transofrm,
+                        'complex64' or 'complex128'.
+        **kwargs        (optional): additional options corresponding to the
+                        entries in the `nufft_opts` structure may be specified
+                        as keyword-only arguments.
+    """
+
+    def __init__(self, nufft_type, n_modes, n_trans=1, eps=1e-6, isign=None,
+                 dtype="complex64", **kwargs):
+        if isign is None:
+            if nufft_type == 2:
+                isign = -1
+            else:
+                isign = +1
+
+        # Need to set the plan here in case something goes wrong later on,
+        # otherwise we error during __del__.
+        self._plan = None
+
+        # Setup type bound methods
+        self.dtype = np.dtype(dtype)
+
+        if self.dtype == np.float64:
+            warnings.warn("Real dtypes are currently deprecated and will be "
+                          "removed in version 2.3. Converting to complex128.",
+                          DeprecationWarning)
+            self.dtype = np.complex128
+
+        if self.dtype == np.float32:
+            warnings.warn("Real dtypes are currently deprecated and will be "
+                          "removed in version 2.3. Converting to complex64.",
+                          DeprecationWarning)
+            self.dtype = np.complex64
+
+        if self.dtype == np.complex128:
+            self._make_plan = _make_plan
+            self._setpts = _set_pts
+            self._exec_plan = _exec_plan
+            self._destroy_plan = _destroy_plan
+            self.real_dtype = np.float64
+        elif self.dtype == np.complex64:
+            self._make_plan = _make_planf
+            self._setpts = _set_ptsf
+            self._exec_plan = _exec_planf
+            self._destroy_plan = _destroy_planf
+            self.real_dtype = np.float32
+        else:
+            raise TypeError("Expected complex64 or complex128.")
+
+        self.dim = len(n_modes)
```

Comment:
Should we handle non-tuples?
`self.dim = 1 if isinstance(n_modes, int) else len(n_modes)`

Reply:
Why not? Added.